### PR TITLE
Migrate to c

### PR DESCRIPTION
**Pull request by mheyer32**

### CODE REVIEW

**Comment by 0xABADCAFE** [comment]
Only 40 files changed? Go go gadget review mode...

File: ab3d2_source/Makefile [hunk 1]
```diff
@@ -1,34 +1,58 @@
...
+VLINK=vlink
+
 STRIP=m68k-amigaos-strip
 
 #find the toolchain's directory
 PREFIX = $(shell ./getprefix.sh "$(CC)")
 
 #-DCD32VER
 
+CFLAGS = -m68020-60 -mtune=68030 -noixemul -g -ggdb -O3
```
Comment (0xABADCAFE) [inline ab3d2_source/Makefile:13]: Does tuning for 030 offer anything 020 doesn't?

File: ab3d2_source/c/draw.c [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+#define VID_FAST_BUFFER_SIZE (SCREEN_WIDTH * SCREEN_HEIGHT + 4095)
+#define PLANESIZE (320 / 8 * 256)
+
+extern void unLHA(REG(a0, void *dst), REG(d0, const void *src), REG(d1, ULONG length), REG(a1, void *workspace),
+                  REG(a2, void *X));
+
+extern const UBYTE draw_BorderPacked_vb[];
+extern ULONG Sys_Workspace_vl[];
+
+static UBYTE draw_Border[320 * 256];
```
Comment (0xABADCAFE) [inline ab3d2_source/c/draw.c:22]: Use defined constants for dimensions here?

File: ab3d2_source/c/draw.c [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    unLHA(Vid_FastBufferPtr_l, draw_BorderPacked_vb, 0, Sys_Workspace_vl, NULL);
+
+    PLANEPTR planes[8];
+
+    for (int p = 0; p < 8; ++p) {
+        planes[p] = Vid_FastBufferPtr_l + PLANESIZE * p;
+    };
+
+    // The image we have has a fixed size
+    PlanarToChunky(draw_Border, planes, 320 * 256);
```
Comment (0xABADCAFE) [inline ab3d2_source/c/draw.c:44]: Use defined constants for dimensions here?

File: ab3d2_source/c/draw.c [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+#include "screen.h"
+#include "system.h"
+
+#include <SDI_compiler.h>
+#include <cybergraphics/cybergraphics.h>
+#include <graphics/gfx.h>
+#include <intuition/intuition.h>
+#include <proto/cybergraphics.h>
+#include <proto/exec.h>
+#include <string.h>  //memset
```
Comment (0xABADCAFE) [inline ab3d2_source/c/draw.c:11]: My inner optimisation nerd will want to revisit the memory copy and fill routines at some point but this is OK for now.

File: ab3d2_source/c/main.c [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#include <stdio.h>
+
+extern void startup(void);
+
+const long __nocommandline=1;
+
+int main(int argc, char *argv[])
+{
+    int rval = 10;
+    if (!sys_OpenLibs())
```
Comment (0xABADCAFE) [inline ab3d2_source/c/main.c:16]: Nitpick: Now that sys_OpenLibs() and sys_CloseLibs() are called externally from the system module, if we are being strict with conventions set before then these should be renamed to Sys_OpenLibs and Sys_CloseLibs.

File: ab3d2_source/c/menu.c [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+fail:
+    Signal(mainTask, SIGBREAKF_CTRL_C);
+    Forbid();
+}
+
+static void DestroyBlitTask(void)
+{
+    LOCAL_SYSBASE();
+
+    KPrintF("DestroyBlitTask\n");
```
Comment (0xABADCAFE) [inline ab3d2_source/c/menu.c:419]: I have some concerns that this will interfere with serial port in two player.

File: ab3d2_source/c/system.c [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+}
+
+void Sys_Done()
+{
+    Draw_Shutdown();
+    sys_RemoveInterrupts();
+    sys_ReleaseHardware();
+}
+
+BOOL sys_OpenLibs(void)
```
Comment (0xABADCAFE) [inline ab3d2_source/c/system.c:107]: Mentioned elsewhere, but we should capitalise this now that it's called from the "outside"...

File: ab3d2_source/c/system.c [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+    CLOSELIB(DOSBase);
+    CLOSELIB(CyberGfxBase);
+}
+
+BOOL sys_InitHardware()
+{
+    LOCAL_SYSBASE();
+    if (SysBase->AttnFlags & AFF_68040) {
+        Sys_Move16_b = ~Sys_Move16_b;
+        Vid_FullScreen_b = ~Vid_FullScreen_b;
```
Comment (0xABADCAFE) [inline ab3d2_source/c/system.c:192]: Does this need changing to Vid_FullScreenTemp_b as per the conversation on EAB?

File: ab3d2_source/c/system.c [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+    }
+
+    if (AllocMiscResource(MR_SERIALPORT, AppName)) {
```
Comment (0xABADCAFE) [inline ab3d2_source/c/system.c:195]: Impossible not to read as "mister serialport"...

**Review by 0xABADCAFE — CHANGES_REQUESTED** [review]
Looking good!
A few minor nitpicks - literals being used where there are defines/constants and the 68040+ fullscreen autoselection.

File: ab3d2_source/Makefile (hunk 1 above)
Comment (mheyer32) [reply ab3d2_source/Makefile:13]: In my experience, the -m68020-60 will actually optimize for 060 (like prefer mul over shifts), but avoid instructions that thee 060 doesn't have (like certain 64bit multiplications/divisions). With -mtune 030 I'm aiming for 030 optimization without using 030 instructions the the 060 doesn't have.

File: ab3d2_source/c/draw.c [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+#define VID_FAST_BUFFER_SIZE (SCREEN_WIDTH * SCREEN_HEIGHT + 4095)
+#define PLANESIZE (320 / 8 * 256)
```
[the remark this answers is not in the document]
Comment (mheyer32) [reply ab3d2_source/c/draw.c:14]: I used these hardcoded values mainly for places where we involve assets that are hardcoded to these sizes. I expect SCREEN_WIDTH and SCREEN_HEIGHT potentially become actual runtime defined some day...

File: ab3d2_source/c/draw.c (hunk 4 above)
Comment (mheyer32) [reply ab3d2_source/c/draw.c:11]: Yeah, this is all non-critical stuff so far. Though the compiler already does a pretty good job at inlining and unrolling memset/memcpy when the size is known at compile time.

File: ab3d2_source/c/main.c (hunk 5 above)
Comment (mheyer32) [reply ab3d2_source/c/main.c:16]: will do

File: ab3d2_source/c/menu.c (hunk 6 above)
Comment (mheyer32) [reply ab3d2_source/c/menu.c:419]: It does. Will have to remove this. We should probably have a debug mode that won't grab the serial port right at startup.

File: ab3d2_source/c/menu.c [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+static void mnu_fade(UWORD fadeFactor);
+
+static CHIP WORD emptySprite[6];
+static struct BitMap mnu_bitmap;
+
+static LONG mnu_subtract;
+static UWORD mnu_count;
+static struct ScreenBuffer *mnu_ScreenBuffer;
+
+struct Task *blitTask;
```
[the remark this answers is not in the document]
Comment (mheyer32) [reply ab3d2_source/c/menu.c:57]: Yeah, I couldn't come up with something better. There was no good entry point to do a mnu_present(). On AGA its mainly event driven (background and fire are interrupt driven).
This is a stopgap solution until we port the menu to chunky rendering. I believe it should be possible to just emulate the fire-blits and the background scroll and the C2P into chipmem on AGA.
But that requires rewriting the event loop for the menu. Keeping the menu be rendered in planar chipmem (even on ECS!) and copying over to chunky via the OS seemed liked the fastest and safest stopgap solution.

File: ab3d2_source/c/draw.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (0xABADCAFE) [reply ab3d2_source/c/draw.c:14]: Having symbolic names for the values is still beneficial, even if we don't use the existing defines. It's a lot easier to replace a named define with confidence than it is a value (speaking from experience with the assembler side already)

**Review by 0xABADCAFE — APPROVED** [review]
Looks good to me!

**Pull request merged**